File: package_patch.py
"""Build-specific, logical UPK patch. Contains no shipped game payload."""
import hashlib
import math
import struct
import lzokay
from bytecode import patch_play_shot

TAG = 0x9E2A83C1
PLAY_OFFSET, PLAY_SIZE = 0xFFC020, 0xF4
PLAY_HASH = 'd1700b21bef4f3b9933e5f5f7382209775d086219defaa44dc9d7d0e510a859d'
CHECK_OFFSET, CHECK_SIZE = 0x12CC5D2, 0x94
CHECK_HASH = '46f89feed0c8bd0577bcca550071a590a29abcf5407ee178a37f600c27ebc7a2'
EXPORT_SLOT = 0x3CF64B


def sha(data):
    return hashlib.sha256(data).hexdigest()


def entries(data):
    if len(data) < 0x75 or struct.unpack_from('<I', data)[0] != TAG:
        raise ValueError('Not a UE3 package')
    if struct.unpack_from('<I', data, 0x6D)[0] != 2:
        raise ValueError('Expected LZO package')
    count = struct.unpack_from('<I', data, 0x71)[0]
    if count != 286 or struct.unpack_from('<I', data, 0x21)[0] != 173666:
        raise ValueError('Unsupported game build (expected 25136512 layout)')
    result = [struct.unpack_from('<IIII', data, 0x75 + i * 16) for i in range(count)]
    last_end = None
    for logical, size, physical, packed in result:
        if size <= 0 or physical + packed > len(data) or packed < 16:
            raise ValueError('Invalid compressed chunk range')
        if last_end is not None and logical != last_end:
            raise ValueError('Non-contiguous logical package')
        last_end = logical + size
    return result


def unpack(data, entry):
    _, expected, offset, packed_size = entry
    tag, block, _, total = struct.unpack_from('<IIII', data, offset)
    if tag != TAG or total != expected or block < 1 or block > 1048576:
        raise ValueError('Invalid LZO header')
    count = math.ceil(total / block)
    cursor = offset + 16 + count * 8
    out = bytearray()
    for index in range(count):
        length, raw_size = struct.unpack_from('<II', data, offset + 16 + index * 8)
        if raw_size != min(block, total - len(out)) or cursor + length > offset + packed_size:
            raise ValueError('Invalid LZO block size')
        payload = data[cursor:cursor + length]
        decoded = payload if length == raw_size else lzokay.decompress(payload, raw_size)
        if len(decoded) != raw_size:
            raise ValueError('Invalid decompressed size')
        out.extend(decoded)
        cursor += length
    if cursor != offset + packed_size:
        raise ValueError('Unexpected compressed trailing data')
    return bytes(out)


def pack(raw, block=131072):
    blocks = []
    for offset in range(0, len(raw), block):
        piece = bytes(raw[offset:offset + block])
        compressed = lzokay.compress(piece)
        blocks.append((compressed if len(compressed) < len(piece) else piece, len(piece)))
    return (struct.pack('<IIII', TAG, block, sum(len(x) for x, _ in blocks), len(raw))
            + b''.join(struct.pack('<II', len(x), n) for x, n in blocks)
            + b''.join(x for x, _ in blocks))


def read_at(data, table, offset, size):
    for index, entry in enumerate(table):
        logical, length, _, _ = entry
        if logical <= offset and offset + size <= logical + length:
            raw = unpack(data, entry)
            return index, raw, raw[offset - logical:offset - logical + size]
    raise ValueError('Requested object crosses chunk boundary')


def build(source):
    table = entries(source)
    _, _, original = read_at(source, table, PLAY_OFFSET, PLAY_SIZE)
    _, _, check = read_at(source, table, CHECK_OFFSET, CHECK_SIZE)
    if sha(original) != PLAY_HASH or sha(check) != CHECK_HASH:
        raise ValueError('Unsupported or modified M2G functions; no files changed')
    first, export_chunk, slot = read_at(source, table, EXPORT_SLOT + 32, 8)
    if slot != struct.pack('<II', PLAY_SIZE, PLAY_OFFSET):
        raise ValueError('PlayShot export already relocated or unsupported')
    replacement = patch_play_shot(original)
    final_index = len(table) - 1
    logical, size, _, _ = table[final_index]
    new_offset = logical + size
    final = unpack(source, table[final_index]) + replacement
    export_chunk = bytearray(export_chunk)
    struct.pack_into('<II', export_chunk, EXPORT_SLOT + 32 - table[first][0], len(replacement), new_offset)
    out = bytearray(source)
    # Append replacement chunks. Existing source bytes, including old PlayShot,
    # remain intact. Only these two directory records point to new chunks.
    for index, raw in [(first, export_chunk), (final_index, final)]:
        compressed = pack(raw)
        physical = len(out)
        out.extend(compressed)
        struct.pack_into('<IIII', out, 0x75 + index * 16, table[index][0], len(raw), physical, len(compressed))
    out = bytes(out)
    updated = entries(out)
    _, _, actual = read_at(out, updated, new_offset, len(replacement))
    if actual != replacement:
        raise ValueError('Patched package verification failed')
    return out


def link_executable(source, before, after):
    if source[:2] != b'MZ':
        raise ValueError('Not a Windows executable')
    needle = b'brggame.upk\0'
    locations = []
    cursor = 0
    while True:
        position = source.find(needle, cursor)
        if position < 0:
            break
        locations.append(position + len(needle))
        cursor = position + len(needle)
    if len(locations) != 2:
        raise ValueError('Unexpected executable manifest entry count')
    old_hash, new_hash = hashlib.sha1(before).digest(), hashlib.sha1(after).digest()
    result = bytearray(source)
    for offset in locations:
        if source[offset:offset + 20] != old_hash:
            raise ValueError('Executable package hash does not match')
        result[offset:offset + 20] = new_hash
    return bytes(result)

File: package-patch.js
'use strict';

const { createHash } = require('node:crypto');
const { lzo1xCompress, lzo1xDecompress } = require('./vendor/lzo1x/dist/index.cjs');
const { patchPlayShot } = require('./bytecode');
const TAG = 0x9e2a83c1;

const LAYOUTS = {
  173666: {
    exportCount: 173666,
    exportSlot: 0x3cf64b,
    playOffset: 0xffc020,
    playSize: 0xf4,
    playHash: 'd1700b21bef4f3b9933e5f5f7382209775d086219defaa44dc9d7d0e510a859d',
    checkOffset: 0x12cc5d2,
    checkSize: 0x94,
    checkHash: '46f89feed0c8bd0577bcca550071a590a29abcf5407ee178a37f600c27ebc7a2',
    playerClassIndex: 53632,
    pawnIndex: 9431,
    atkTypeIndex: 9325,
    commonPawnIndex: 9428,
    setBulletName: 62614,
    playAnimName: 57693,
    reflectSkillName: 59818,
  },
  173671: {
    exportCount: 173671,
    exportSlot: 0x3cf700,
    playOffset: 0xffc23d,
    playSize: 0xf4,
    playHash: 'b81b4b96b0d20c15d3a66a49b7ce2da140e902d3eabb4d58f0cc18453cb799cc',
    checkOffset: 0x12cc85e,
    checkSize: 0x94,
    checkHash: 'e9671b9d3c7a065025f96ecace016cfa034216477ac9d2072e88c9448a93e94f',
    playerClassIndex: 53637,
    pawnIndex: 9431,
    atkTypeIndex: 9325,
    commonPawnIndex: 9428,
    setBulletName: 62616,
    playAnimName: 57695,
    reflectSkillName: 59820,
  }
};

const sha = data => createHash('sha256').update(data).digest('hex');

function words(...values) {
  const b = Buffer.alloc(values.length * 4);
  values.forEach((n, i) => b.writeUInt32LE(n, i * 4));
  return b;
}

function getLayout(data) {
  const count = data.readUInt32LE(0x71);
  const exportCount = data.readUInt32LE(0x21);
  if (count !== 286 || !LAYOUTS[exportCount]) {
    throw new Error(`Unsupported game build (chunk count ${count}, export count ${exportCount})`);
  }
  return LAYOUTS[exportCount];
}

function entries(data) {
  if (data.length < 0x75 || data.readUInt32LE() !== TAG) throw new Error('Not a UE3 package');
  if (data.readUInt32LE(0x6d) !== 2) throw new Error('Expected LZO package');
  getLayout(data);
  const count = data.readUInt32LE(0x71);
  if (data.length < 0x75 + count * 16) throw new Error('Truncated chunk directory');
  const result = []; let lastEnd;
  for (let i = 0; i < count; i++) {
    const entry = Array.from({ length: 4 }, (_, n) => data.readUInt32LE(0x75 + i * 16 + n * 4));
    const [logical, size, physical, packed] = entry;
    if (!size || size > 64 * 1024 * 1024 || physical < 0x75 + count * 16 || physical + packed > data.length || packed < 16 || logical + size > 0xffffffff) throw new Error('Invalid compressed chunk range');
    if (lastEnd !== undefined && logical !== lastEnd) throw new Error('Non-contiguous logical package');
    lastEnd = logical + size; result.push(entry);
  }
  return result;
}

function unpack(data, entry) {
  const [, expected, offset, packedSize] = entry;
  if (offset < 0 || packedSize < 16 || offset + packedSize > data.length) throw new Error('Invalid LZO range');
  const tag = data.readUInt32LE(offset), block = data.readUInt32LE(offset + 4);
  const compressedTotal = data.readUInt32LE(offset + 8), total = data.readUInt32LE(offset + 12);
  if (tag !== TAG || total !== expected || total > 64 * 1024 * 1024 || block < 1 || block > 1048576) throw new Error('Invalid LZO header');
  const count = Math.ceil(total / block);
  let cursor = offset + 16 + count * 8, written = 0, payloadTotal = 0;
  if (cursor > offset + packedSize) throw new Error('Truncated LZO block table');
  const out = Buffer.alloc(total);
  for (let i = 0; i < count; i++) {
    const length = data.readUInt32LE(offset + 16 + i * 8), rawSize = data.readUInt32LE(offset + 20 + i * 8);
    if (!length || rawSize !== Math.min(block, total - written) || cursor + length > offset + packedSize) throw new Error('Invalid LZO block size');
    const payload = data.subarray(cursor, cursor + length);
    const decoded = length === rawSize ? payload : lzo1xDecompress(payload, rawSize);
    if (decoded.length !== rawSize) throw new Error('Invalid decompressed size');
    out.set(decoded, written); written += rawSize; cursor += length; payloadTotal += length;
  }
  if (cursor !== offset + packedSize || payloadTotal !== compressedTotal) throw new Error('Unexpected compressed size/trailing data');
  return out;
}

function pack(raw, block = 131072) {
  if (!Number.isInteger(block) || block < 1 || block > 131072) throw new Error('Invalid output block size');
  const blocks = [], sizes = [];
  for (let offset = 0; offset < raw.length; offset += block) {
    const piece = raw.subarray(offset, offset + block), compressed = Buffer.from(lzo1xCompress(piece));
    const payload = compressed.length < piece.length ? compressed : piece;
    if (payload !== piece && !Buffer.from(lzo1xDecompress(payload, piece.length)).equals(piece)) throw new Error('LZO compression verification failed');
    blocks.push(payload); sizes.push(words(payload.length, piece.length));
  }
  return Buffer.concat([words(TAG, block, blocks.reduce((n, b) => n + b.length, 0), raw.length), ...sizes, ...blocks]);
}

function readAt(data, table, offset, size) {
  for (let index = 0; index < table.length; index++) {
    const [logical, length] = table[index];
    if (logical <= offset && offset + size <= logical + length) {
      const raw = unpack(data, table[index]);
      return { index, raw, data: raw.subarray(offset - logical, offset - logical + size) };
    }
  }
  throw new Error('Requested object crosses chunk boundary');
}

function build(source) {
  const layout = getLayout(source);
  const table = entries(source), original = readAt(source, table, layout.playOffset, layout.playSize).data;
  const check = readAt(source, table, layout.checkOffset, layout.checkSize).data;
  if (sha(original) !== layout.playHash || sha(check) !== layout.checkHash) throw new Error('지원되지 않거나 변경된 M2G 함수입니다. 파일을 변경하지 않았습니다.');
  const exp = readAt(source, table, layout.exportSlot + 32, 8);
  if (!exp.data.equals(words(layout.playSize, layout.playOffset))) throw new Error('이미 적용됐거나 지원되지 않는 PlayShot 위치입니다.');
  const replacement = patchPlayShot(original, layout), last = table.length - 1;
  const newOffset = table[last][0] + table[last][1];
  const final = Buffer.concat([unpack(source, table[last]), replacement]);
  words(replacement.length, newOffset).copy(exp.raw, layout.exportSlot + 32 - table[exp.index][0]);
  const out = Buffer.from(source), append = []; let physical = out.length;
  for (const [index, raw] of [[exp.index, exp.raw], [last, final]]) {
    const compressed = pack(raw);
    words(table[index][0], raw.length, physical, compressed.length).copy(out, 0x75 + index * 16);
    append.push(compressed); physical += compressed.length;
  }
  const result = Buffer.concat([out, ...append]);
  if (!readAt(result, entries(result), newOffset, replacement.length).data.equals(replacement)) throw new Error('Patched package verification failed');
  return result;
}

function linkExecutable(source, before, after) {
  if (source.subarray(0, 2).toString('ascii') !== 'MZ') throw new Error('Not a Windows executable');
  const needle = Buffer.from('brggame.upk\0'), locations = []; let cursor = 0;
  while (true) {
    const at = source.indexOf(needle, cursor);
    if (at < 0) break;
    cursor = at + needle.length; locations.push(cursor);
  }
  if (locations.length !== 2) throw new Error('Unexpected executable manifest entry count');
  const oldHash = createHash('sha1').update(before).digest(), newHash = createHash('sha1').update(after).digest();
  const result = Buffer.from(source);
  for (const offset of locations) {
    if (!source.subarray(offset, offset + 20).equals(oldHash)) throw new Error('Executable package hash does not match');
    newHash.copy(result, offset);
  }
  return result;
}

// Keep 25136512 constants for backwards compatibility with existing tests
module.exports = {
  sha, entries, unpack, pack, readAt, build, linkExecutable,
  PLAY_OFFSET: LAYOUTS[173666].playOffset,
  PLAY_SIZE: LAYOUTS[173666].playSize,
  EXPORT_SLOT: LAYOUTS[173666].exportSlot,
  LAYOUTS, getLayout
};
